add k8s object receiver (#588)
signalfx/splunk-otel-collector-chart: UPGRADING.md

# Upgrade guidelines

## 0.64.0 to 0.66.0
Before this change, [k8s events receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/k8seventsreceiver) was used to collect Kubernetes Events.
Now we utilize [k8s objects receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/k8sobjectsreceiver) that can pull or watch any object from Kubernetes API server.

> Therefore, `clusterReceiver.eventsEnabled` is now deprecated, and to maintain the same behavior you should set

Reviewer: https://stackoverflow.com/a/9208164

It appears these changes remove `clusterReceiver.eventsEnabled` functionality in total, making it obsolete. Is this correct?

Author: Yes, right, k8s object receiver is not providing exactly the same functionality as k8s events receiver. Updated.

Reviewer: So https://github.com/signalfx/splunk-otel-collector-chart/pull/588#discussion_r1036766723 could also be expanded so that we provide an actual deprecation path by using the objects receiver w/ just the watched events for this value instead of obsoleting in one pass. The warning check also needs to be updated with a description of the behavior.

imo users should never have their config become "broken" during an upgrade and we should provide at least a release of warnings and ideally remediation logic in helpers (cc @dmitryax)

Author: Ok, so I updated it in the way that when `eventsEnabled` is set to true we return warning message and event objects are watched, so behaviour is similar (but not the same, as previous event formatting was different, and the sourcetype was `kube:events` and now it is `kube:object<resource_id>`. It is described in details in the upgrading doc.

# Upgrade guidelines

> ## 0.64.0 to 0.66.0

Reviewer: Would be 0.67.0 if part but maybe a version placeholder makes more sense.

Author: Changed it to `$CURRENT_VERSION to $NEXT_VERSION`. Same in NOTES.txt

Before this change, [k8s events receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/k8seventsreceiver) was used to collect Kubernetes Events.

> Now we utilize [k8s objects receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/k8sobjectsreceiver) that can pull or watch any object from Kubernetes API server.

Reviewer: ```suggestion
Now we utilize the [Kubernetes Objects Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/k8sobjectsreceiver) that can pull or watch any object from Kubernetes API server.
```

Therefore, `clusterReceiver.eventsEnabled` is now deprecated - when on, it applies the following receiver config:

```yaml
k8sobjects:
  auth_type: serviceAccount
  objects:
    - mode: watch
      name: events
```

Which is an equivalent of gathering events by k8s events receiver.

To achieve the same behavior using `clusterReceiver.objectsEnabled` parameter, set:

```yaml
objectsEnabled: true
k8sObjects:
  - mode: watch
    name: events
```

> There is a difference in the event formatting between `k8s_events` receiver and `k8sobjects` receiver results.
> `k8s_events` receiver stores event message as a log body, with the following fields added as a resource/attribute:
> * k8s.object.kind
> * k8s.object.name
> * k8s.object.uid
> * k8s.object.fieldpath
> * k8s.object.api_version
> * k8s.object.resource_version
> * k8s.event.reason
> * k8s.event.action
> * k8s.event.start_time
> * k8s.event.name
> * k8s.event.uid
> * k8s.namespace.name

Reviewer: ```suggestion
There are differences in the log record formatting of events between the previous `k8s_events` receiver and the now adopted `k8sobjects` receiver. The `k8s_events` receiver stores event messages their log body, with the following fields added as attributes:

* `k8s.object.kind`
* `k8s.object.name`
* `k8s.object.uid`
* `k8s.object.fieldpath`
* `k8s.object.api_version`
* `k8s.object.resource_version`
* `k8s.event.reason`
* `k8s.event.action`
* `k8s.event.start_time`
* `k8s.event.name`
* `k8s.event.uid`
* `k8s.namespace.name`
```

> In case of `k8sobjects`, the whole payload is stored in the log body and `object.message` refers to the event message.

Reviewer: ```suggestion
Now with the `k8sobjects` receiver, the whole payload is stored in the log body and `object.message` refers to the event message.
```

Mind showing an example message?

> You can monitor more kubernetes objects configuring `clusterReceiver.k8sObjects` according to the instruction from
> [k8s objects receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/k8sobjectsreceiver).

Reviewer: ```suggestion
You can monitor more Kubernetes objects configuring by `clusterReceiver.k8sObjects` according to the instructions from the
[Kubernetes Objects Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/k8sobjectsreceiver) documentation.
```
